File: 2023/day05/go/src/seedmap/seedmap.go
package seedmap

import "fmt"

type SeedMap struct {
	SourceStart uint64
	SourceEnd   uint64
	TargetStart uint64
	TargetEnd   uint64
	Size        uint64
}

func New(source uint64, target uint64, size uint64) *SeedMap {
	sm := new(SeedMap)
	sm.SourceStart = source
	sm.SourceEnd = source + size
	sm.TargetStart = target
	sm.TargetEnd = target + size
	sm.Size = size

	return sm
}

func NewFromString(seedMapString string) *SeedMap {
	sm := new(SeedMap)

	return sm
}

func (sm *SeedMap) MapSeed(value *uint64) bool {
	fmt.Println("Mapping seed")

	if (*value >= sm.SourceStart) && (*value < sm.SourceEnd) {
		*value = *value - sm.SourceStart + sm.TargetStart
		return true
	}

	return false
}

File: 2023/day05/c_sharp/app_code/SeedMap.cs


namespace seed_map
{
    public class SeedMap
    {
        UInt64 source_;
        UInt64 target_;
        UInt64 size_;

        public UInt64 Source { get=> source_;}
        public UInt64 Target { get=> target_;}
        public UInt64 Size { get=> size_;}

        public SeedMap(UInt64 source, UInt64 target, UInt64 size)
        {
            source_ = source;
            target_ = target;
            size_ = size;
        }

        public SeedMap(string seed_map_string)
        {
            var seed_strings = seed_map_string.Split(" ");
            if (seed_strings.Length != 3)
            {
                throw new Exception("Can't parse the line");
            }

            source_ = UInt64.Parse(seed_strings[1]);
            target_ = UInt64.Parse(seed_strings[0]);
            size_ = UInt64.Parse(seed_strings[2]);
        }

        public UInt64? MapSeed(UInt64 seed)
        {
            if (
                (source_ <= seed) && 
                ((source_ + size_) > seed)
            )
            {
                return seed - source_ + target_;
            }

            return null;
        }
    }
}